idaes/core/initialization/general_hierarchical.py

```diff
@@ -0,0 +1,253 @@
+
+    def _prepare_addons(self, model, addon_args, logger):
+        sub_initializers = {}
+        addon_args = dict(addon_args)
+        for sm in model.initialization_order:
+            if sm is not model:
+                # Get initializers for add-ons
+                sub_initializers[sm] = self.get_submodel_initializer(sm)()
+
+                if sm not in addon_args.keys():
+                    addon_args[sm] = {}
+
+                # Call prepare method for add-ons
+                sub_initializers[sm].addon_prepare(sm, **addon_args[sm])
+
+        logger.info_high("Step 1: preparation complete.")
+
+        return sub_initializers, addon_args
+
+    def _initialize_submodels(
+        self, model, addon_args, copy_inlet_state, sub_initializers, logger
+    ):
+        results = None
+
+        for sm in model.initialization_order:
+            if sm is model:
+                results = self._initialize_main_model(model, copy_inlet_state, logger)
+            else:
+                sub_initializers[sm].addon_initialize(sm, **addon_args[sm])
+
+        if results is None:
+            raise InitializationError(
+                f"Main model ({model.name}) was not initialized (no results returned). "
+                f"This is likely due to an error in the model.initialization_order."
+            )
+
+        logger.info_high("Step 2: sub-model initialization complete.")
+
+        return results
+
+    def _initialize_main_model(self, model, copy_inlet_state, logger):
+        # Initialize properties
+        try:
+            # Guess a 0-D control volume
+            self._init_props_0D(model, copy_inlet_state)
+        except AttributeError:
+            # Assume it must be a 1-D control volume
+            self._init_props_1D(model)
+
+        logger.info_high("Step 2a: properties initialization complete")
+
+        # Initialize reactions if they exist
+        if hasattr(model.control_volume, "reactions"):
+            self._init_rxns(model)
+        logger.info_high("Step 2b: reactions initialization complete")
+
+        # Solve main model
+        solve_log = idaeslog.getSolveLogger(
+            model.name, self.config.output_level, tag="unit"
+        )
+
+        with idaeslog.solver_log(solve_log, idaeslog.DEBUG) as slc:
+            results = self._get_solver().solve(model, tee=slc.tee)
+
+        logger.info_high(f"Step 2c: model {idaeslog.condition(results)}.")
+
+        return results
+
+    def _init_props_0D(self, model, copy_inlet_state):
+        # Initialize inlet properties - inlet state should already be fixed
+        prop_init = self.get_submodel_initializer(model.control_volume.properties_in)
+
+        if prop_init is not None:
+            prop_init.initialize(
+                model.control_volume.properties_in,
+                solver=self.config.solver,
+                optarg=self.config.solver_options,
+                outlvl=self.config.output_level,
+            )
+
+        if not copy_inlet_state:
```

Comment:
This is because the `D version only has 1 state block (which is indexed between 0 and 1), and thus when we call the Initializer for `control_volume.properties` it will end up initializing all elements (we might be able to build a "copy" state" into the state block initializers to reduce the problem size there, but that is not something we should do at the unit model level).